wallet/src/main/java/com/android/identity_credential/wallet/WalletApplication.kt

```diff
@@ -0,0 +1,284 @@
+import android.graphics.Rect
+import android.graphics.Shader
+import android.widget.Toast
+import com.android.identity.android.securearea.AndroidKeystoreSecureArea
+import com.android.identity.android.storage.AndroidStorageEngine
+import com.android.identity.credential.CredentialStore
+import com.android.identity.credential.NameSpacedData
+import com.android.identity.internal.Util
+import com.android.identity.mdoc.mso.MobileSecurityObjectGenerator
+import com.android.identity.mdoc.mso.StaticAuthDataGenerator
+import com.android.identity.mdoc.util.MdocUtil
+import com.android.identity.securearea.SecureArea
+import com.android.identity.securearea.SecureAreaRepository
+import com.android.identity.util.Logger
+import com.android.identity.util.Timestamp
+import org.bouncycastle.asn1.x500.X500Name
+import org.bouncycastle.cert.X509CertificateHolder
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter
+import org.bouncycastle.cert.jcajce.JcaX509v3CertificateBuilder
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import org.bouncycastle.operator.ContentSigner
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.math.BigInteger
+import java.security.KeyPair
+import java.security.KeyPairGenerator
+import java.security.SecureRandom
+import java.security.Security
+import java.security.cert.X509Certificate
+import java.security.spec.ECGenParameterSpec
+import java.util.Date
+import java.util.Random
+import kotlin.math.ceil
+
+class WalletApplication : Application() {
+    companion object {
+        private const val TAG = "WalletApplication"
+
+        val MDL_DOCTYPE = "org.iso.18013.5.1.mDL"
+        val MDL_NAMESPACE = "org.iso.18013.5.1"
+        val AAMVA_NAMESPACE = "org.iso.18013.5.1.aamva"
+    }
+
+    lateinit var secureAreaRepository: SecureAreaRepository
+    lateinit var credentialStore: CredentialStore
+
+    private lateinit var androidKeystoreSecureArea: AndroidKeystoreSecureArea
+
+    override fun onCreate() {
+        super.onCreate()
+        Logger.d(TAG, "onCreate")
+
+        // This is needed to prefer BouncyCastle bundled with the app instead of the Conscrypt
+        // based implementation included in the OS itself.
+        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME)
+        Security.addProvider(BouncyCastleProvider())
+
+        // Setup singletons
+        val storageDir = File(applicationContext.noBackupFilesDir, "identity")
+        val storageEngine = AndroidStorageEngine.Builder(applicationContext, storageDir).build()
+        secureAreaRepository = SecureAreaRepository()
+        androidKeystoreSecureArea = AndroidKeystoreSecureArea(applicationContext, storageEngine)
+        secureAreaRepository.addImplementation(androidKeystoreSecureArea);
+        credentialStore = CredentialStore(storageEngine, secureAreaRepository)
+    }
+
+    private fun createArtwork(color1: Int,
```

Comment:
we can decouple these and all provision-related code into a provisioning-specific file/class for better modularity and encapsulation of related business logic

Reply:
Yep, exactly!